### PR TITLE
[4.0] Moving com_finder submenu to mainmenu

### PR DESCRIPTION
**Pull request by Hackwar**
This adds menu item types for com_finder views, removes the submenu and adds them into the mainmenu. 

### How to test
1. Apply patch
2. Go to database view and click on "fix"
3. Go to menu manager -> All menu items and click "rebuild"
4. See that the Smart Search menu item can now open up and contains 4 sub menu items and that the submenu in the actual views has been removed.

### CODE REVIEW

**Comment by brianteeman** [comment]
Please alpha sort the language file and check to see if we need the _DESC - most of those are completely unnecessary and we have in general removed all _DESC from J4

**Comment by Hackwar** [comment]
Fixed the language string. The DESC string is maybe not necessary, but it is still used everywhere (100+ results if you search for it) so if you want to remove those, please do that consistently in another PR.

**Comment by brianteeman** [comment]
@Hackwar It has already been done. The only _DESC strings that remain are the few that we are still displaying and the _DESC strings that are for DESCENDING not DESCRIPTION

**Comment by Hackwar** [comment]
No, it hasn't been done, because these strings are fresh from the backend menu items of com_content.

**Comment by brianteeman** [comment]
image

**Comment by Hackwar** [comment]
https://github.com/joomla/joomla-cms/pull/24020/files#diff-1cea33e0f389319a3d2adc85965a3e72R5

**Comment by Hackwar** [comment]
There are currently 40 default.xml files with a DESC string in them in /administrator/components, excluding the 4 that I've added in this PR. As I said, if that is not wanted/allowed anymore, then please open a PR that removes all these.

**Comment by brianteeman** [comment]
Sorry I didnt realise they were for the menu descriptions. Those are being kept - its the _desc that were used as tooltips that have gone .

/me needs coffee

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 544ace46a0450aa3f19e144cdfaa2b2cbed963f8<br /><br />Followed test instructions but didnt get a sub menu<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 8ac75818cde92a8d7c43a8220824309712e8e02d<br /><br />tested OK on a clean install


screen shot 2019-03-04 at 17 11 30
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

File: administrator/components/com_finder/Helper/FinderHelper.php [hunk 1]
```diff
@@ -26,39 +26,6 @@ class FinderHelper
...
-	 * @param   string  $vName  The name of the active view.
-	 *
-	 * @return  void
-	 *
-	 * @since   2.5
-	 */
-	public static function addSubmenu($vName)
-	{
-		\JHtmlSidebar::addEntry(
-			\JText::_('COM_FINDER_SUBMENU_INDEX'),
```
Comment (Quy) [inline administrator/components/com_finder/Helper/FinderHelper.php]: Delete the 4 obsolete strings.

**Comment by infograf768** [comment]
@Hackwar 
Also please solve conflicts.

**Comment by Hackwar** [comment]
I can't see the error that I have in the postgres SQL file. Could someone have a look here?

**Comment by Hackwar** [comment]
Tests have passed now.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on e525eb7f79b0b921bbfe4351fb48a507e71e4271<br /><br />OK with new install.
No submenus when updating using Patchtester, but it might be an issue with Patchtester. Manually ran update SQL successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on e525eb7f79b0b921bbfe4351fb48a507e71e4271<br /><br />Tested successfully with new installation.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

File: administrator/components/com_finder/Helper/FinderHelper.php [hunk 2]
```diff
@@ -29,39 +29,6 @@ class FinderHelper
...
-	 * @param   string  $vName  The name of the active view.
-	 *
-	 * @return  void
-	 *
-	 * @since   2.5
-	 */
-	public static function addSubmenu($vName)
-	{
-		\JHtmlSidebar::addEntry(
-			Text::_('COM_FINDER_SUBMENU_INDEX'),
```
Comment (Quy) [inline administrator/components/com_finder/Helper/FinderHelper.php:44]: Remove `use Joomla\CMS\Language\Text;`

File: administrator/components/com_menus/Model/ItemModel.php [hunk 3]
```diff
@@ -1360,6 +1360,8 @@ public function save($data)
 	{
 		$pk         = (!empty($data['id'])) ? $data['id'] : (int) $this->getState('item.id');
 		$isNew      = true;
+		$db      = $this->getDbo();
```
Comment (Quy) [inline administrator/components/com_menus/Model/ItemModel.php]: This file should not be in this PR.

**Comment by brianteeman** [comment]
@franz-wohlkoenig please removed RTC for now
@hackwar can you please address the two comments from @quy

Thanks

**Comment by Hackwar** [comment]
done

**Comment by brianteeman** [comment]
Database fix will not find the updates as they have an old date. Changing the date to current should resolve that

There is a copy paste error in the sql. You have not updated the "class:associations"

From looking at others this should be "class:finder", "class:finder-maps" etc

Also I'm not sure what the best practice is but the update sql does not have the lft/rght values

**Comment by brianteeman** [comment]
Can you look at resolving the conflicts so that this can be tested please.

**Comment by richard67** [comment]
@brianteeman Your comment above about database fixer and file names is wrong. Database fix will not do anything with the schema updates, because they contain only DML (data manipulation language like INSERT, UPDATE, DELETE). The database fixed only handles DDL (data definition language like CREATE TABLE, ALTER TABLE, DROP TABLE). This does not depend on the file name. The file name is only relevant for version compare when updating, so the installer runs only schema updates with version newer than schema version in database before the update.

@Hackwar You should update your testing instructions accordingly. Fix button does not make sense. Schema updates have to be applied e.g. in PhpMyAdmin or PhpPgAdmin or so.

**Comment by brianteeman** [comment]
@richard67 the file name means that the database fix wont even see the update. Whether it works or not is a different issue.

**Comment by richard67** [comment]
@brianteeman You are wrong. The database fixer always checks all available schema update files, if there are some from 2.5 it will be checked, too, even when being on a 3.9.10. It is the installer/updater where the file name is relevant. I always wanted to change that, but it would be a bigger development changing the weak points of the database fix, and so I haven't made it yet. But of course you are right with the file names having to be fixed. Only your understanding on the database fixer is wrong.

**Comment by brianteeman** [comment]
learnt something new

File: administrator/components/com_finder/View/Filters/HtmlView.php [hunk 4]
```diff
@@ -107,8 +98,6 @@ public function display($tpl = null)
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
 
-		FinderHelper::addSubmenu('filters');
```
Comment (Quy) [inline administrator/components/com_finder/View/Filters/HtmlView.php:110]: Remove `use Joomla\Component\Finder\Administrator\Helper\FinderHelper;`

File: administrator/components/com_finder/View/Maps/HtmlView.php [hunk 5]
```diff
@@ -111,8 +102,6 @@ public function display($tpl = null)
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
 
-		FinderHelper::addSubmenu('maps');
```
Comment (Quy) [inline administrator/components/com_finder/View/Maps/HtmlView.php:114]: Remove `use Joomla\Component\Finder\Administrator\Helper\FinderHelper;`

File: administrator/components/com_finder/View/Searches/HtmlView.php [hunk 6]
```diff
@@ -102,8 +102,6 @@ public function display($tpl = null)
 			throw new GenericDataException(implode("\n", $errors), 500);
 		}
 
-		FinderHelper::addSubmenu('searches');
```
Comment (Quy) [inline administrator/components/com_finder/View/Searches/HtmlView.php:105]: Remove `use Joomla\Component\Finder\Administrator\Helper\FinderHelper;`

File: administrator/components/com_finder/View/Filters/HtmlView.php (hunk 4 above)
Comment (Hackwar) [reply administrator/components/com_finder/View/Filters/HtmlView.php:110]: Done

File: administrator/components/com_finder/View/Maps/HtmlView.php (hunk 5 above)
Comment (Hackwar) [reply administrator/components/com_finder/View/Maps/HtmlView.php:114]: Done

File: administrator/components/com_finder/View/Searches/HtmlView.php (hunk 6 above)
Comment (Hackwar) [reply administrator/components/com_finder/View/Searches/HtmlView.php:105]: Done

**Comment by Hackwar** [comment]
Could those who tested this earlier, please test again, so that we can merge this?

**Comment by richard67** [comment]
@Hackwar You should adjust your testing instructions. The Database Fix will not run the insert statements in the schema update because it only handles DDL, as I mentioned in in my comment above. People either have to do a clean install or they have to run the insert statements from the schema update in PhpMyAdmin (or PgMyAdmin if PostgreSQL) after having applied the patch.

**Comment by brianteeman** [comment]
Please remove the composer.lock file

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 822d13b5b9441603f00da91b702bacbf11a8d2d1<br /><br />Work ok on a clean install
Does not work on an update - no submenus - and yes I did paste the sql into phpmyadmin<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by bees4ever** [comment]
Unfortunaltely it is not wokring for me.

Submenu in the actual views has been removed but there are no 4 sub items in the menu.

Bildschirmfoto 2019-09-02 um 11 52 00

**Comment by bees4ever** [comment]
I have tested this item :red_circle: unsuccessfully on 7e9fac5e9523b20a39cd5351eef0fb962cc23fbf<br /><br />Unfortunaltely it is not wokring for me.

see https://github.com/joomla/joomla-cms/pull/24020#issuecomment-527085371<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by roland-d** [comment]
@Hackwar For me this also doesn't work. The submenu items are gone after applying the patch but the  Fix database and Menu Rebuild doesn't add it to the submenu.

**Comment by roland-d** [comment]
I have tested this item :red_circle: unsuccessfully on 7e9fac5e9523b20a39cd5351eef0fb962cc23fbf<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by brianteeman** [comment]
@roland-d can you or one of the students confirm my finding that it was ok on a clean install.  Also re your comment about database fix see https://github.com/joomla/joomla-cms/pull/24020#issuecomment-523620049

**Comment by roland-d** [comment]
@brianteeman The test done by @bees4ever was on a clean installation as in he cloned the repo, installed J4 and did the test.

**Comment by brianteeman** [comment]
@roland-d can you or the students test #26137

**Comment by nadjak77** [comment]
I have tested this item :red_circle: unsuccessfully on 7e9fac5e9523b20a39cd5351eef0fb962cc23fbf<br /><br />not sure if I understand the right way.
point 4: if I do a new menu item "smart search" there is still only one submenu<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Comment by Quy** [comment]
Please test PR #27032<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24020">issues.joomla.org/tracker/joomla-cms/24020</a>.</sub>

**Pull request closed without merging**